### PR TITLE
 Refactor string call handling and tighten string method validation

### PR DESCRIPTION
**Pull request by sergillam**
Refactor Python frontend string call handling to reduce duplicated argument parsing logic.
Add shared helpers for positional and keyword argument resolution.
Tighten validation for string method arguments and error paths.
Fix scope counting for nested function, class, and lambda bodies.
Add regression tests for nondeterministic string containment with embedded null characters.

### CODE REVIEW

**Review by brcfarias — CHANGES_REQUESTED** [review]
Could you keep each PR focused on a single change? This one mixes several things.

File: src/python-frontend/string_handler.cpp [hunk 1]
```diff
@@ -88,6 +96,898 @@ format_value_from_json(const nlohmann::json &arg, python_converter &converter)
...
+      {
+        return converter.get_expr(receiver_json["func"]["value"]);
+      }
+    }
+    return nil_exprt();
+  }
+
+  if (method_name == "join")
+  {
+    ensure_allowed_keywords(method_name, keyword_values, {});
```
[the remark this answers is not in the document]
Comment (sergillam) [reply src/python-frontend/string_handler.cpp:848]: done

File: src/python-frontend/string_handler.cpp [hunk 2]
```diff
@@ -679,11 +1579,31 @@ std::string string_handler::ensure_string_function_symbol(
...
+{
+  auto cache_it = symbol_cache_.find(symbol_id);
+  if (cache_it != symbol_cache_.end())
+    return cache_it->second;
+
+  symbolt *symbol = symbol_table_.find_symbol(symbol_id);
+  if (symbol != nullptr)
+    symbol_cache_.emplace(symbol_id, symbol);
+
+  return symbol;
```
[the remark this answers is not in the document]
Comment (sergillam) [reply src/python-frontend/string_handler.cpp:1600]: done

File: src/python-frontend/string_handler_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+  for (const auto &kw : keywords)
+  {
+    if (!kw.contains("arg") || !kw["arg"].is_string() || !kw.contains("value"))
+      continue;
+
+    const std::string keyword_name = kw["arg"].get<std::string>();
+    if (!keyword_values.emplace(keyword_name, &kw["value"]).second)
+    {
+      throw std::runtime_error(
```
[the remark this answers is not in the document]
Comment (sergillam) [reply src/python-frontend/string_handler_utils.cpp]: done

File: src/python-frontend/string_handler_utils.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  const nlohmann::json &keywords);
+
+const nlohmann::json *find_keyword_value(
+  const keyword_valuest &keyword_values,
+  const std::string &name);
+
+void ensure_allowed_keywords(
+  const std::string &method_name,
+  const keyword_valuest &keyword_values,
+  std::initializer_list<const char *> allowed);
```
[the remark this answers is not in the document]
Comment (sergillam) [reply src/python-frontend/string_handler_utils.h:28]: done

**Review by lucasccordeiro — APPROVED** [review]
Thanks for submitting this PR, @sergillam.
